R/plumber-step.R

```diff
@@ -61,6 +61,13 @@ PlumberStep <- R6Class(
         self$serializer <- serializer
       }
     },
+    #' @description convert args to specific types
+    #' @param args List of args to convert
+    convertArgs = function(args) {
+      # a general step doesn't convert args
+      # but a subclass can choose to override and convert
```

Comment:
Clever. But I think I'd like to do this earlier in the process. I'll see if I can find a better location. (It definitely won't be cleaner than this.)

Reply:
No opinion from me - this was just where I had the args and the params conveniently together.

I think you have permissions to edit this branch and contents - so please do :+1: - I also won't be overly offended if you through everything away and do it better :+1:

Reply:
Notes for me....

This method is only being applied to the handle's non-`req`/`res` arguments. It will not apply to `pr$argsPath`, `pr$argsQuery` or any other stored values.  Just the functions helper parameters that have typically been discouraged.

Since we'll have the location of `query` and `path` in the parameter definitions, we should process the parameters as soon as possible (such as in the [query filter](https://github.com/rstudio/plumber/blob/76cecc58925897c91cb88203ab485a0c512a0d26/R/parse-query.R#L6) and [path param extraction](https://github.com/rstudio/plumber/blob/76cecc58925897c91cb88203ab485a0c512a0d26/R/plumber.R#L689)).

This method should just be a helper method that is applied to the args given.

... however... the query filter does not know which endpoint is going to be executed. This makes typing the arguments impossible (or not something we should do). 

-----------------

This still feels weird to me, but we're being caught by the query parameters being stored on `pr$args` too early.

I'm worried about having a path parameter being parsed as an integer and the user adding a matching parameter query string value that is not an integer. Ex: `/iteration/5?i=A` where the route is defined as `@param i:int iteration`. This would set `req$args` to `list(i = "A", i = 5)` and `req$args[["i"]]` would resolve to `"A"`. 

This is just a problem in general about which param is returned. So maybe it's no worse that what it currently is.

--------------------

Therefore, this **is** the correct location to parse the args and only apply them to the helper params (at the cost of possibly overlapping which parameter is actually retrieved).

Reply:
One possibility here is to deprecate putting the type in the routes - suggest all type information is placed in the @param

Obviously that's not pain free for existing code, but should be possible to keep support and to just add a warning that this will get switched off at some point in the future.

But this is probably easier for me to say as I don't have lots of existing plumber APIs to maintain and update.

Reply:
@slodge I like the idea of deprecating the type in the routes, we could tie this new behavior to an edition.